### DISCUSSION TITLE
How to achieve bitwise logic composition inside FastAPI dependencies

### DISCUSSION THREAD

**Discussion by 7xMind** [fastapi/fastapi#8341] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# My Desired Outcome

@router.get("/test-auth", tags=["Internal"])
def my_api(has_perm: bool = Depends(CheckPerms(perm_classes=[IsAdmin() | HasPerm('write')]))-> dict:
    response.status_code = status.HTTP_200_OK
    perm = perm
    return {"status": perm}

# Here's what could be behind Perm classes:

    class Auth:
        @abstractmethod
        def has_perm(request_header: Header):
            ...
    
        def __or__(self, other):
            return self.has_perm(request_header) or other.has_perm(request_header)
    
    class IsAdmin(Auth):
        def has_perm(request_header):
            # Dummy
            return request_header['is_admin'] is True
    
    class HasPerm(Auth):
        def has_perm(request_header):
            return x in request_header
```

### Description

In Django rest framework there's this nice feature where you can compose permission classes with bitwise operations as such:

```python
class MyApi(APIView):
    permission_calsses = [IsAdmin | HasPerm]
```

I want to achieve a similar behavior in FastAPI aiming at such an interface:

```python
@router.get("/test-auth", tags=["Internal"])
def my_api(has_perm: bool = Depends(CheckPerms(perm_classes=[IsAdmin() | HasPerm('write')]))-> dict:
    response.status_code = status.HTTP_200_OK
    perm = perm
    return {"status": perm}

```

My problem is that this expression `perm_classes=[IsAdmin() | HasPerm('write')]` is evaluated as soon as I run the server. But I need to find a way to pass some arguments to these perm_classes when the request arrives and postpone the evaluation of `[IsAdmin | HasPerm]` to when the request has arrived and I can pass it to these classes. 

Any ideas?

# Here's what could be behind Perm classes:

```python
class Auth:
    @abstractmethod
    def has_perm(request_header: Header):
        ...

    def __or__(self, other):
        return self.has_perm(request_header) or other.has_perm(request_header)

class IsAdmin(Auth):
    def has_perm(request_header):
        # Dummy
        return request_header['is_admin'] is True

class HasPerm(Auth):
    def has_perm(request_header):
        return x in request_header
```

### Operating System

macOS

### FastAPI Version

0.78

### Python Version

3.8.12

### Additional Context

None

**Comment by JarroVGIT** [comment]
You could add the request to the __call__ method of `CheckPerms` and to the and to subsequent other classes , so when it is called upon dependency resolution, the request is passed through to the other classes as well.

**Comment by 7xMind** [comment]
> You could add the request to the **call** method of `CheckPerms` and to the and to subsequent other classes , so when it is called upon dependency resolution, the request is passed through to the other classes as well.

You are correct. But the problem is that the bitwise operation like `or` will call the `has_perm(request_header)` on subclasses **before** the `__call__` on `CheckPerm` is called. Isn't it so?

**Comment by JarroVGIT** [comment, marked as answer]
I think I see your problem now. I don't know how you could solve that. I thinkered around a bit, and tried something like the below. Is that helpful at all? It seems to me it is overly complex and I am not sure what the benefit is..

```python
from tabnanny import check
from typing import Any, Callable
from fastapi import FastAPI, Depends, HTTPException, Request
from functools import partial
app = FastAPI()

#note, this cannot be a class. Classes that are dependencies are initiated on startup, 
#so you can't pass the request object (as that is not available, at startup).
def checkPerms(or_classes: list[list[Callable | str]], request: Request) -> bool:
    result = []
    #let's call all subclasses
    for cls in or_classes:
        #check if there are arguments
        if len(cls) > 1:
            c = cls[0]()
            res = c(*cls[1::], request)
        else:
            #else call the class directly without arguments.
            c = cls[0]()
            res = c(request)
        if isinstance(res, bool):
            result.append(res)
        else:
            raise HTTPException(detail="Authentication did not return bool")
        #return if any of the results is True.
    return sum(result) > 0

class IsAdmin:
    def __init__(self) -> None:
        return

    def __call__(self, request: Request) -> bool:
        #simulation: if there is a ?q=whatever present in the url, this will return True.
        if request.query_params.get("q"):
            return True
        return False

class HasPerm:
    def __init__(self) -> None:
        return

    def __call__(self, permission: str, request: Request) -> bool:
        #just some logic, to illustrate how it works with params.
        if not permission:
            return False
        if permission == "read":
            return True
        if permission == "write":
            return False

checkperms_admin_write = partial(checkPerms, [[IsAdmin], [HasPerm, "write"]])
checkperms_admin_read = partial(checkPerms, [[IsAdmin], [HasPerm, "read"]])
@app.get("/with-write-perm")
async def with_write(perms:bool = Depends(checkperms_admin_write)):
    #returns false without ?q= in url
    return perms
    
@app.get("/with-read-perm")
async def with_read(perms:bool = Depends(checkperms_admin_read)):
    return perms


if __name__ == "__main__":
    import uvicorn
    uvicorn.run(app, host="0.0.0.0", port=8000,  )
```
